antenna/ext/gcs/crashstorage.py

```diff
@@ -0,0 +1,163 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import logging
+import os
+import uuid
+
+from everett.manager import Option
+from google.auth.credentials import AnonymousCredentials
+from google.cloud import storage
+
+from antenna.app import register_for_verification
+from antenna.ext.crashstorage_base import CrashStorageBase
+from antenna.util import json_ordered_dumps
+
+logger = logging.getLogger(__name__)
+
+
+def generate_test_filepath():
+    """Generate a unique-ish test filepath."""
+    return "test/testfile-%s.txt" % uuid.uuid4()
+
+
+class GcsCrashStorage(CrashStorageBase):
+    """Save raw crash files to GCS.
+
+    This will save raw crash files to GCS in a pseudo-tree something like this:
+
+    ::
+
+        <BUCKET>
+           v1/
+               dump_names/
+                   <CRASHID>
+               <DUMPNAME>/
+                   <CRASHID>
+               raw_crash/
+                   <YYYYMMDD>/
+                       <CRASHID>
+
+
+    **Authentication**
+
+    The google cloud sdk will automatically detect credentials as described in
+    https://googleapis.dev/python/google-api-core/latest/auth.html:
+
+    - If you're running in a Google Virtual Machine Environment (Compute Engine, App
+      Engine, Cloud Run, Cloud Functions), authentication should "just work".
+    - If you're developing locally, the easiest way to authenticate is using the `Google
+      Cloud SDK <http://cloud.google.com/sdk>`_::
+
+        $ gcloud auth application-default login
+
+    - If you're running your application elsewhere, you should download a `service account
+      <https://cloud.google.com/iam/docs/creating-managing-service-accounts#creating>`_
+      JSON keyfile and point to it using an environment variable::
+
+        $ export GOOGLE_APPLICATION_CREDENTIALS="/path/to/keyfile.json"
+
+
+    **Local emulator**
+
+    If you set the environment variable ``STORAGE_EMULATOR_HOST=http://host:port``,
+    then this will connect to a local GCS emulator.
+
+
+    """
+
+    class Config:
+        bucket_name = Option(
+            doc=(
+                "Google Cloud Storage bucket to save to. Note that the bucket must "
+                "already have been created."
+            ),
+        )
+
+    def __init__(self, config):
+        self.config = config.with_options(self)
+        self.bucket = self.config("bucket_name")
+
+        if emulator := os.environ.get("STORAGE_EMULATOR_HOST"):
+            logger.debug(
+                "STORAGE_EMULATOR_HOST detected, connecting to emulator: %s",
+                emulator,
+            )
+            self.client = storage.Client(
+                credentials=AnonymousCredentials(),
+                project="test",
+            )
+        else:
+            self.client = storage.Client()
```

Comment:
unlike for pubsub in https://github.com/mozilla-services/antenna/pull/974#discussion_r1497952874 the default retry and timeout behavior for google-cloud-storage is better defined. we use two network methods in this class: `Client.get_bucket` and `Blob.upload_from_string`, both of which have a default timeout of 60 seconds.  `Blob.upload_from_string` is expected not to retry because the client assumes it is not an idempotent action. `Client.get_bucket` may retry and I think that's fine given there is a default timeout set.

tl;dr I think this client sets sane defaults for retry and timeout, and we shouldn't mess with them unless/until we see an issue.